tool/cargo/Syncer.kt

```diff
@@ -0,0 +1,407 @@
+
+                data class Crate(override val name: String, val version: String, val features: List<String>) : Dependency(name) {
+                    override fun toToml(bazelOutputBase: File): Config {
+                        return Config.inMemory().apply {
+                            set<String>("version", version)
+                            set<List<String>>("features", features)
+                        }
+                    }
+                }
+
+                data class Path(override val name: String, val path: String) : Dependency(name) {
+                    override fun toToml(bazelOutputBase: File): Config {
+                        return Config.inMemory().apply {
+                            set<String>("path", path.replace(Paths.EXTERNAL_PLACEHOLDER, bazelOutputBase.resolve(Paths.EXTERNAL).absolutePath))
+                        }
+                    }
+                }
+
+                companion object {
+                    fun of(rawKey: String, rawValue: String): Dependency {
+                        val name = rawKey.split(".", limit = 2)[1]
+                        val rawValueProps = rawValue.split(";")
+                                .associate { it.split("=", limit = 2).let { parts -> parts[0] to parts[1] } }
+                        return if (Keys.VERSION in rawValueProps) {
+                            Crate(
+                                    name = name,
+                                    version = rawValueProps[Keys.VERSION]!!,
+                                    features = rawValueProps[Keys.FEATURES]?.split(",") ?: emptyList()
+                            )
+                        } else {
+                            Path(name = name, path = rawValueProps[Keys.PATH]!!)
+                        }
+                    }
+                }
+            }
+
+            enum class Type {
+                LIB,
+                BIN,
+                TEST,
+                BUILD;
+
+                companion object {
+                    fun of(value: String): Type {
+                        return when (value) {
+                            "lib" -> LIB
+                            "bin" -> BIN
+                            "test" -> TEST
+                            "build" -> BUILD
+                            else -> throw IllegalArgumentException()
+                        }
+                    }
+                }
+            }
+
+            companion object {
+                fun fromPropertiesFile(path: Path): TargetProperties {
+                    val props = Properties().apply { load(FileInputStream(path.toString())) }
+                    try {
+                        return TargetProperties(
+                                path = path,
+                                name = props.getProperty(Keys.NAME),
+                                label = props.getProperty(Keys.LABEL),
+                                type = Type.of(props.getProperty(Keys.TYPE)),
```

Comment:
note: removed some of the static imports that were not obvious to the the reader when new to the project -- this is preference though

Reply:
My reasoning for statically importing `NAME`, `LABEL` etc. is that we always strive for canonical names (this is a principle discussed extensively in the Rust book, too!), and in this case it is quite clear that `props.getProperty(NAME)` is reading some kind of "name" property from the `Properties` object. `Keys` is redundant because, clearly, we're reading a key because that's what the method `getProperty` does.

Reply:
Alright, I've reverted this to be just using `NAME`, etc. directly :) Makes sense!